ons-spark/raw-notebooks/groups-not-loops/groups-not-loops.ipynb

```diff
@@ -0,0 +1,1166 @@
+       "9   040568-06042016     2016      652.0   \n",
+       "10  050586-29042016     2016      326.0   \n",
+       "11  099706-31072016     2016      326.0   \n",
+       "\n",
+       "                                          Description AnimalGroup  \n",
+       "0    CAT STUCK WITHIN WALL SPACE  RSPCA IN ATTENDANCE         Cat  \n",
+       "1                           KITTEN TRAPPED IN CHIMNEY         Cat  \n",
+       "2                    ASSIST RSPCA WITH CAT IN CHIMNEY         Cat  \n",
+       "3                DEER IN CANAL - WATER RESCUE LEVEL 2        Deer  \n",
+       "4         DEER ENTANGLED - RSPCA ON SCENE 07969305977        Deer  \n",
+       "5   TWO DEER STUCK IN FENCING RVP OUTSIDE PORTLAND...        Deer  \n",
+       "6                       DOG TRAPPED BETWEEN TWO HOUSE         Dog  \n",
+       "7   PUPPY TRAPPED BETWEEN TWO HOUSES  FRU REQ FROM...         Dog  \n",
+       "8   DOG STUCK ON ISLAND IN LARGE POND WATER RESCUE...         Dog  \n",
+       "9                     HAMSTER STUCK UNDER FLOORBOARDS     Hamster  \n",
+       "10                       HAMSTER STUCK IN CAVITY WALL     Hamster  \n",
+       "11                  HAMSTER TRAPPED UNDER FLOORBOARDS     Hamster  "
+      ]
+     },
+     "execution_count": 8,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "def get_expensive_incidents_grouped(df, animals):\n",
+    "    return(df\n",
+    "        .filter(F.col(\"AnimalGroup\").isin(animals)) # Only include specified animals\n",
+    "        .select(\"IncidentNumber\", \"CalYear\", \"TotalCost\", \"Description\", \"AnimalGroup\") # Add AnimalGroup to the select\n",
+    "        .withColumn(\"row_no\", F.row_number().over(Window # Rank the animals\n",
+    "                                                .partitionBy(\"AnimalGroup\") # Group by AnimalGroup\n",
+    "                                                .orderBy(F.desc(\"TotalCost\"), \"IncidentNumber\"))) # Order as before\n",
+    "        .filter(F.col(\"row_no\") <= 3) # Only return the top 3 per group\n",
+    "        .drop(\"row_no\") # Remove the column used in the calculation\n",
+    "          )\n",
+    "\n",
+    "(get_expensive_incidents_grouped(rescue, animals)\n",
+    "    .orderBy([\"AnimalGroup\", \"TotalCost\"], ascending=[True, False])\n",
+    "    .toPandas()\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "```r\n",
+    "get_expensive_incidents_grouped <- function(df, animals){\n",
+    "  df %>%\n",
+    "    sparklyr::filter(AnimalGroup %in% animals) %>%\n",
+    "    sparklyr::select(IncidentNumber, CalYear, TotalCost, Description, AnimalGroup) %>% \n",
+    "    dplyr::arrange(desc(TotalCost), AnimalGroup, IncidentNumber) %>%\n",
+    "    sparklyr::mutate(row_no = row_number()) %>%\n",
+    "    sparklyr::filter(row_no <= 3) %>%\n",
+    "    sparklyr::select(-row_no)\n",
```

Comment:
I could not run the notebook converter.  

The `sparklyr::select(-row_no)` threw an error(My `dplyr` version is `1.1.4’), there is nothing wrong with the code please refer here https://github.com/tidyverse/dbplyr/issues/1245.  

If i take off the code line which caused the error the notebook converter worked fine. There are some issues with python and R tabs not being together. I will send the output html files separately since GitHub doesn't support them.

Reply:
Hi Anton, thanks for reviewing and flagging those issues. I will ensure the Python and R code is together, didn't see that line there between the two cells! 

Will removing the desc from "dplyr::arrange(desc(TotalCost), AnimalGroup, IncidentNumber) %>%" resolve the issue? If so, I can remove it

Reply:
Unfortunately it didn't resolve it, how about replacing
`sparklyr::mutate(row_no = row_number()) %>%`
`sparklyr::filter(row_no <= 3) %>%`
`sparklyr::select(-row_no)`

with `head(3)` as you did in other parts?

Reply:
Hi Anton, head(3) would only give the first 3 results of the df, but it is the top 3 most expensive results from each animal group that are needed (hence the row_no <= 3). I spotted an error in the code, it was missing the group_by and ungroup to make the window function work properly. I've fixed it and pushed the changes. Can you check if that will convert now?

Reply:
Oh makes sense , thanks for explaining, the error still persists maybe somebody else with a newer version of dplyr can convert it ( i will try to upgrade it too)

Alternative how about not creating a row_number column and filter it directly to the row_number() like this :

get_expensive_incidents_grouped <- function(df, animals){
  df %>%
    sparklyr::filter(AnimalGroup %in% animals) %>%
    sparklyr::select(IncidentNumber, CalYear, TotalCost, Description, AnimalGroup) %>% 
    dplyr::group_by(AnimalGroup) %>%
    dplyr::arrange(desc(TotalCost)) %>%
    sparklyr::filter(row_number() <= 3) %>%
    dplyr::arrange(AnimalGroup,desc(TotalCost), IncidentNumber) 
}

Reply:
Hey Johanna,

Everything looks fine, thank you. One final thing,  we also need to add the `groups-not-loops.md` in the concepts folder , which is basically just running the converter . Are you able to do that ? I remember you had some issues with running  the converter , happy to upload it for you.

Reply:
Hi Anton,
Still unable to run the converter unfortunately :( Would you be able to run it to get the groups-not-loops.md and upload it to the concepts folder please?

Reply:
Just uploaded a commit with the converter outputs : 

- `converter.py` has  just paths for groups not loops and targeting the concepts folder
- `ons-spark/raw-notebooks/groups-not-loops` has `outputs.csv` and `r_input.R`  
- `ons-spark/spark-concepts/groups-not-loops.md` the converter output 

I believe we are good to merge :)